### PR TITLE
version history

### LINKED ISSUE #7: Add version history to edited documents
[Fronted Expectations](http://upload.montague.im/u/d7vlTf.png)
We must ensure this is mobile friendly.

For the backend the following endpoints should/may be added
- [x] **GET** `/document/{documentID}/versions`
  - Returns an array of all document versions sorted version
- [x] **GET** `/document/{documentID}/versions/{version}`
  - Return a documents for a specific version
- [x] **GET** `/raw/{documentID}/versions/{version}`
- [x] **DELETE** `/documents/{documentID}/versions/{version}`
  - Deletes a specific version from the document

The following endpoints should be updated
- [x] **POST** `/documents/`
  - Should include a version 
- [x] **DELETE** `/documents/{documentID}/`
  - Deletes all document versions
- [x] **PATCH** `/documents/{documentID}/`
  - Should create a new version of the document with the new content

The following database changes should/may be made
- [x] new `version` column on the `documents` table
- [x] update primary key to include `id` & `version`
- [ ] remove `created_at` column

### PR DESCRIPTION
**Pull request by day-mon**

### CODE REVIEW

File: assets/script.js [hunk 1]
```diff
@@ -59,27 +59,26 @@ document.addEventListener("keydown", (event) => {
...
+            doKeyboardAction(event, "edit")
             break;
         case "d":
-            event.preventDefault();
-            if (document.querySelector("#delete").disabled) return;
-            document.querySelector("#delete").click();
+            doKeyboardAction(event, "delete")
             break;
     }
 })
```
Comment (Sanduhr32) [inline assets/script.js]: extract the string and call doKeyboardAction() once

**Review by Sanduhr32 — COMMENTED** [review]
funny

File: routes.go [hunk 2]
```diff
@@ -55,16 +56,20 @@ func (s *Server) Routes() http.Handler {
...
 	r.Head("/documents/{documentID}", s.GetDocument)
 	r.Get("/documents/{documentID}", s.GetDocument)
 	r.Patch("/documents/{documentID}", s.PatchDocument)
 	r.Delete("/documents/{documentID}", s.DeleteDocument)
+	r.Delete("/documents/{documentID}/versions/{version}", s.DeleteDocumentByVersion)
 
 	r.Get("/{documentID}", s.GetPrettyDocument)
 	r.Head("/{documentID}", s.GetPrettyDocument)
 
+	r.Get("/document/{documentID}/versions", s.DocumentVersions)
```
Comment (topi314) [inline routes.go]: should be `/documents/{documentID}/versions`

File: routes.go [hunk 3]
```diff
@@ -80,6 +85,86 @@ func (s *Server) Assets() http.Handler {
...
+			Key:      version.ID,
+			Version:  version.Version,
+			Data:     version.Content,
+			Language: version.Language,
+		})
+	}
+	s.JSON(w, r, response)
+}
+
+func (s *Server) DeleteDocumentByVersion(w http.ResponseWriter, r *http.Request) {
```
Comment (topi314) [inline routes.go]: should be `DeleteDocumentVersion`

File: routes.go [hunk 4]
```diff
@@ -80,6 +85,86 @@ func (s *Server) Assets() http.Handler {
...
+		s.Error(w, r, err, http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/plain; charset=UTF-8")
+	w.WriteHeader(http.StatusOK)
+	_, _ = w.Write([]byte(document.Content))
+}
+
+func sanitizeDocumentWithVersion(r *http.Request, s *Server, w http.ResponseWriter) (string, int64) {
```
Comment (topi314) [inline routes.go]: should be `parseDocumentVersion`

File: database.go [hunk 5]
```diff
@@ -124,12 +171,13 @@ func (d *Database) createDocument(ctx context.Context, content string, language
...
 	doc := Document{
 		ID:          id,
+		Version:     time.Now().Unix(),
 		Content:     content,
 		Language:    language,
 		UpdateToken: updateToken,
-		UpdatedAt:   time.Now(),
 	}
-	res, err := d.NamedExecContext(ctx, "UPDATE documents SET content = :content, language = :language, updated_at = :updated_at WHERE id = :id AND update_token = :update_token", doc)
+	// TODO:this should do an insert with a new version now
```
Comment (topi314) [inline database.go]: this to do is done

File: database.go [hunk 6]
```diff
@@ -87,6 +102,39 @@ func (d *Database) GetDocument(ctx context.Context, id string) (Document, error)
...
+}
+
+func (d *Database) GetDocumentByVersion(ctx context.Context, id string, version int64) (Document, error) {
+	var doc Document
+	err := d.GetContext(ctx, &doc, "SELECT * FROM documents WHERE id = $1 AND version = $2", id, version)
+	return doc, err
+}
+
+func (d *Database) DeleteDocumentByVersion(ctx context.Context, version int64, documentID string) error {
+	res, err := d.ExecContext(ctx, "DELETE FROM documents WHERE version = $1 AND id = $2", version, documentID)
```
Comment (topi314) [inline database.go]: should be `DeleteDocumentByVersion(ctx context.Context, documentID string, version int64)`

File: database.go [hunk 7]
```diff
@@ -87,6 +102,39 @@ func (d *Database) GetDocument(ctx context.Context, id string) (Document, error)
...
+	if withContent {
+		sqlString = "SELECT id, version, content, language FROM documents where id = $1 ORDER BY version DESC"
+	} else {
+		sqlString = "SELECT id, version FROM documents where id = $1 ORDER BY version DESC"
+	}
+	err := d.SelectContext(ctx, &docs, sqlString, id)
+	return docs, err
+}
+
+func (d *Database) GetDocumentByVersion(ctx context.Context, id string, version int64) (Document, error) {
```
Comment (topi314) [inline database.go]: should be `GetDocumentByVersion(ctx context.Context, documentID string, version int64)`

File: database.go [hunk 8]
```diff
@@ -87,6 +102,39 @@ func (d *Database) GetDocument(ctx context.Context, id string) (Document, error)
 	return doc, err
 }
 
+func (d *Database) GetDocumentVersions(ctx context.Context, id string, withContent bool) ([]Document, error) {
```
Comment (topi314) [inline database.go]: should be `GetDocumentVersions(ctx context.Context, documentID string, withContent bool)`

**Review by topi314 — CHANGES_REQUESTED** [review]

**Review by topi314 — APPROVED** [review]

**Pull request merged**